cesium/internal/domain/file_controller.go

```diff
@@ -332,14 +330,30 @@ func (fc *fileController) removeReadersWriters(ctx context.Context, key uint16)
 }
 
 func (fc *fileController) close() error {
-	fc.writers.Lock()
-	defer fc.writers.Unlock()
+	fc.writers.RLock()
+	fc.readers.RLock()
+	defer func() {
+		fc.readers.RUnlock()
+		fc.writers.RUnlock()
+	}()
 	c := errors.NewCatcher(errors.WithAggregation())
 	for _, w := range fc.writers.open {
+		c.Exec(func() error {
+			if !w.tryAcquire() {
+				return errors.Newf("writer for file %d is in use and cannot be closed", w.fileKey)
```

Comment:
Might be good to figure out a way to inject the channel key and name before we pass this up the callstack. Would be good to know exactly which db still has files open.

Reply:
All failures at this level must go through the unary level, so is it fine if this error only passes the mesasge, and it will be wrapped later?